[PNP-5602] Remove slimmer/static from Frontend

sairamya93 opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Remove dependency on `static` service calls via slimmer for layouts

- Remove slimmer gem
- Layouts now directly call the `layout_for_public` component
- `Full_width` layout for homepage, landing page, service manual and service toolkit pages
- Emergency/Global banner support provided by the `govuk_web_banners` gem

## Why

As part of [RFC176 - Replacing Static](https://github.com/alphagov/govuk-rfcs/pull/176), we replace the static service by moving asset handling into frontend and using the `govuk_web_banners` gem for banners. This improves maintainability and enables easier deployments.

[Trello card](https://trello.com/c/qWE04tVs/778-remove-slimmer-frontend), [Jira issue PNP-5602](https://gov-uk.atlassian.net/browse/PNP-5602)

## Example pages

- [Homepage](https://govuk-frontend-app-pr-4928.herokuapp.com/)
- [Service Manual](https://govuk-frontend-app-pr-4928.herokuapp.com/service-manual)
- [Service Toolkit](https://govuk-frontend-app-pr-4928.herokuapp.com/service-toolkit)
- [Service Manual topic](https://govuk-frontend-app-pr-4928.herokuapp.com/service-manual/agile-delivery)
- [Answer page](https://govuk-frontend-app-pr-4928.herokuapp.com/national-minimum-wage-rates)
- [Speech](https://govuk-frontend-app-pr-4928.herokuapp.com/government/speeches/spending-review-2025-speech)

sairamya93 committed:
Remove slimmer gem and config

- Removed slimmer gem and configuration
- Updated controllers that relied on slimmer layout
- Removed and updated tests that expected slimmer behaviour

sairamya93 committed:
Add basic elements bundle

- As done in email-alert-frontend - https://github.com/alphagov/email-alert-frontend/pull/1890, bundled everything used in all pages of the app here
- Added basic component stylesheets in application.scss
- anything in application.scss added to the custom exclude list so that it isn't included twice by the components gem.
- Added the JS files needed for those components (previously supplied by static)

sairamya93 committed:
Update application layout with direct component calls

-  As in email-alert-frontend, calling stylesheet_link_tag "application" to optimize contents of the link header, which allows browsers to download CSS early before rendering
- Updated layout to use layout_for_public component directly
- Called render_component_stylesheets as part of the content_for :head, which is yielded to in the layout
- Meta tag rendering and banners functionality in subsequent commits
Ref. link: https://github.com/alphagov/email-alert-frontend/pull/1890/commits/da605fbbca4a8f12e915e0aa614818765d3b959e
- Helper stylesheet_link_tag is added as part of layout_for_public rendering and hence is not required in application layout

Co-authored-by: Keith Lawrence <KludgeKML@users.noreply.github.com>
Co-authored-by: Andy Sellick <andysellick@users.noreply.github.com>

sairamya93 committed:
Add global/emergency banner in application layout

- Added global/emergency banner in layout as done for email-alert-frontend and feedback
Ref. link: https://github.com/alphagov/email-alert-frontend/pull/1890/commits/b08add52f18a609284af92acdb7b3d4b42ace549
The related banner JS dependencies have been already added
- Added initializer for emergency banner REDIS support
- Added mock redis client for request and system tests

sairamya93 committed:
Add full width layout for specific pages

- Added 'full_width' layout for below pages as the layout for these pages vary from the default 'application' layout:
  Homepage
  Landing page
  Service manual index
  Service toolkit

- Added homepage flag to be true as the margins for homepage in mobile and tablets are wider compared to rest of the pages.

sairamya93 committed:
Add breadcrumbs to application layout

Ref. from original file: https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L10-L12

sairamya93 committed:
Add phase banner

Ref. link to the original file:
https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L7-L9

sairamya93 committed:
Add page title to application layout

- Added the page_title by refactoring the existing code
Ref. link: https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L30

sairamya93 committed:
Update extra_headers to head in all the files

As done in [feedback](https://github.com/alphagov/feedback/pull/2108/commits/4734e7a2eebc898052ca9e98a1be1df94e753191), updating extra_headers to head in all the rendering page views

sairamya93 committed:
Add html escaping in local transaction tests with invalid postcode

- To ensure correct rendering of error messages having a single quote('), updated the tests to use HTML-escaping

sairamya93 committed:
Remove wrapper class method

- As we don't have slimmer anymore for frontend, the wrapper id and class in the original file are not added to the layout
Ref. link: https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L63
- Also removed wrapper class method in Application helper and related test

sairamya93 committed:
Add javscript tags in application layout

- Added javascript tags for test-specific JS and extra_javascript to include when required by individual pages
- JS bundle for the application is added through layout_for_public - https://github.com/alphagov/govuk_publishing_components/blob/d4fad2c49d02c06f52464735cd9357f53e55ee67/app/views/govuk_publishing_components/components/_layout_for_public.html.erb#L131

Ref. from original file: https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L48-L51

sairamya93 committed:
Update default_font_face flag in application stylesheet

- Changed 'govuk-include-default-font-face' flag to 'true' as now Static is no longer used to provide the font.

sairamya93 committed:
Move javascript tag in index page for travel advice

- Moved JS tag to the end of the file to ensure it gets executed after the whole page is loaded. Otherwise the filtering in travel_advice page fails to work.

sairamya93 committed:
Fix: breadcrumbs in service manual

deborahchua commented on app/views/layouts/application.html.erb:
```diff
@@ -27,9 +27,13 @@
-
+<%
+  page_title = yield :title
+  content_title = page_title(content_item) || "GOV.UK"
```
I think we can also combine L32-L33 as
`page_title = page_title.presence || page_title(content_item) || "GOV.UK"`

andysellick reviewed (commented):
All looks very clean, good job. My only comments are very minor ones.

I've not finished going through all the pages and visually checking them but my comment above shows progress, will update.

AshGDS commented on app/views/travel_advice/index.html.erb:
```diff
@@ -92,9 +91,9 @@
     <meta name="description" content="<%= @presenter.description %>">
   <% end %>
 <% end %>
+<%= javascript_include_tag "views/travel-advice.js", integrity: false %>
```
@andysellick Having it in the head somehow now causes an issue finding the filter element in the DOM, so I told her to put it at the end of the `<body>`. I think maybe `slimmer` did something smart to make the JS be able to find the DOM element? 

I did try and modify the JS to run in a `DOMContentLoaded` event, but there's a type of Rails test that executes JS that I haven't encountered that started failing, as I don't think that test fires `DOMContentLoaded` when it renders a page (maybe we could mock the event?)

I didn't want to block her work trying to modify the JS, so I thought just moving the file to load after DOM element renders would be OK for now. I think ideally that JS should be converted to a GOVUK module as well, as it isn't using our standard way of adding JS to an element.

sairamya93 commented on app/views/layouts/application.html.erb:
```diff
@@ -1,67 +1,46 @@
+<% stylesheet_link_tag "application", media: "all"  # Trigger this without rendering it to ensure it appears in links header %>
+<% emergency_banner = render("govuk_web_banners/emergency_banner") %>
+<% global_banner = render("govuk_web_banners/global_banner") %>
 <% content_for :body do %>
```
Yes. As mentioned in the [govuk_publishing_components doc](https://github.com/alphagov/govuk_publishing_components/blob/main/docs/set-up-individual-component-css-loading.md#hoist-body-content-to-the-top-of-your-layouts), it is required for the components in the body to be seen before `render_component_stylesheets`, to render the styles correctly.
Thanks @AshGDS for clarifying that.

sairamya93 committed:
Add OpenSearch description link to layouts

- As in the original file, added OpenSearch link to the application and full_width layout files
https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L40

sairamya93 committed:
Add meta tags in layouts

- Added meta tags to full_width and application layout as in the original file
Ref. link: https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L53-L59

sairamya93 committed:
feat: add middleware to optimise GA4 data attributes

- Added middleware in application config to wrap GA4 data attributes in single quote which was previously done by Slimmer
Ref link: https://github.com/alphagov/static/blob/main/docs/moving-an-app-off-static.md#8-add-middleware-to-optimise-ga4-data-attributes

andysellick commented on app/views/layouts/full_width.html.erb:
```diff
@@ -0,0 +1,28 @@
+    emergency_banner:,
+    global_banner:,
+    full_width: true,
+    homepage: true,
```
This value for `homepage` needs to only be `true` for the homepage. This is because on the homepage (on mobile and tablet) the left/right margins are wider than all other pages on GOV.UK, so this flag exists to tell the component to indent slightly on mobile.

This is what's causing the crown to be further across on full width pages on mobile:

Current | This PR
------ | ------
<img width="563" height="70" alt="Screenshot 2025-09-18 at 15 08 40" src="https://github.com/user-attachments/assets/93b1f4f4-3711-44d6-a5f3-172b1f92ee6d" /> | <img width="563" height="73" alt="Screenshot 2025-09-18 at 15 08 47" src="https://github.com/user-attachments/assets/be0ae097-67fe-4d83-9598-01844db02622" />

There's probably a few ways to fix this. One could be to set a variable in the [homepage controller](https://github.com/alphagov/frontend/pull/4928/files#diff-df45d28d309e062c2ac48c98ccba67637a3d94a815b1f246ec42174962c05351) e.g. `homepage_layout = true` that is then used in this call to the layout component e.g. `homepage: homepage_layout,`.

Thanks to @MartinJJones for explaining how all this works.

sairamya93 committed:
feat: Add health check endpoint for emergency banner redis

- Added readiness health check endpoint to check the health of the Emergenecy banner redis
Ref. link: https://github.com/alphagov/static/blob/main/docs/moving-an-app-off-static.md#10-add-healthchecks-for-the-govuk_web_banners-redis

sairamya93 committed:
fix: Remove custom CSS in popular links

- This is removed as it is no longer required to handle the icon size

sairamya93 committed:
fix: Update action_link component call in homepage

- Added 'large' option to the action_link as in here - https://components.publishing.service.gov.uk/component-guide/action_link/large that will fix the action link size for the popular links in the homepage

AshGDS reviewed (approved):
LGTM 👍

sairamya93 committed:
feat: Add recruitment banner in application layout

- Added in recruitment banner as in existing application layout
Ref. https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L13
